algorithms/rainbow/main.py

```diff
@@ -0,0 +1,177 @@
+                               args.history_length, args.device)
+        args.resolution = env.env.config['resolution']
+    else:
+        env = Env(args)
+        env.train()
+        args.resolution = (84, 84)
+    action_space = env.action_space
+
+    # Agent
+    dqn = Agent(args, env)
+    mem = ReplayMemory(args, args.memory_capacity)
+    priority_weight_increase = (1 - args.priority_weight) / (args.T_max - args.learn_start)
+
+    # Construct validation memory
+    val_mem = ReplayMemory(args, args.evaluation_size)
+    T, done = 0, True
+    for T in range(args.evaluation_size):
+        if done:
+            state, done = env.reset(), False
+        next_state, _, done, _ = env.step(env.action_space.sample())
+        val_mem.append(state, None, None, done)
+        state = next_state
+
+    if args.evaluate:
+        dqn.eval()  # Set DQN (online network) to evaluation mode
+        avg_reward, avg_Q = test(env, T, args, dqn, val_mem, evaluate=True)
+        print('Avg. reward: ' + str(avg_reward) + ' | Avg. Q: ' + str(avg_Q))
+    else:
+        # Training loop
+        dqn.train()
+        T, done = 0, True
+
+        while T < args.T_max:
+            if T % 200 == 0:
+                print("step {}".format(T))
+
+            if done:
+                state, done = env.reset(), False
+            if T % args.replay_frequency == 0:
+                dqn.reset_noise()  # Draw a new set of noisy epsilons
+
+            action = dqn.act(state)  # Choose an action greedily (with noisy weights)
+            next_state, reward, done, _ = env.step(action)  # Step
+            if args.reward_clip > 0:
+                reward = max(min(reward, args.reward_clip), -args.reward_clip)  # Clip rewards
+            mem.append(state, action, reward, done)  # Append transition to memory
+            T += 1
+
+            if T % args.log_interval == 0:
+                log('T = ' + str(T) + ' / ' + str(args.T_max))
+
+            # Train and test
+            if T >= args.learn_start:
+                # Anneal importance sampling weight β to 1
+                mem.priority_weight = min(mem.priority_weight + priority_weight_increase, 1)
+
+                if T % args.replay_frequency == 0:
+                    dqn.learn(mem)  # Train with n-step distributional double-Q learning
+
+                if T % args.evaluation_interval == 0:
+                    dqn.eval()  # Set DQN (online network) to evaluation mode
+                    avg_reward, avg_Q = test(env, T, args, dqn, val_mem)
+                    log('T = ' + str(T) + ' / ' + str(args.T_max) + ' | Avg. reward: ' +
+                        str(avg_reward) + ' | Avg. Q: ' + str(avg_Q))
+                    dqn.train()  # Set DQN (online network) back to training mode
+
+                # Update target network
+                if T % args.target_update == 0:
+                    dqn.update_target_net()
+            state = next_state
+    env.close()
```

Comment:
does this work in atari too?
I saw elsewhere you had a check if it was ai2thor and then did this

Reply:
Yes this works well. In order for atari to work, we need to create an env for training and another for evaluation and that's why I close only atari env in test function. I will document that better